### PR TITLE
Verify dashboard API endpoint

### LINKED ISSUE #125: Tye applications running inside a devcontainers aren't visible in the TreeView
Issue Type: <b>Bug</b>

1. Create a devcontainer enabled Tye application. (I tested on the microservices-accelerator app)
2. Run the application using the `tye-run` task.
3. Browse to https://localhost:8000 to see if the app has started.
4. Check if the TreeView is updated with the started services.

Extension version: 0.3.0-alpha
VS Code version: Code - Insiders 1.59.0-insider (599f5d2c67571f84a045a59eb3423bc254ee87d9, 2021-07-16T09:56:58.707Z)
OS version: Windows_NT x64 10.0.22000
Restricted Mode: No
Remote OS version: Linux x64 5.10.43.3-microsoft-standard-WSL2

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i7-8700 CPU @ 3.20GHz (12 x 3192)|
|GPU Status|2d_canvas: enabled<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>oop_rasterization: enabled<br>opengl: enabled_on<br>rasterization: enabled<br>skia_renderer: enabled_on<br>video_decode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|undefined|
|Memory (System)|31.85GB (13.99GB free)|
|Process Argv|--crash-reporter-id 5efa374b-afb0-449b-a378-e88461e67ccf|
|Screen Reader|no|
|VM|0%|

|Item|Value|
|---|---|
|Remote|Dev Container: .NET Core 3.1 with Docker, Dapr, and Tye|
|OS|Linux x64 5.10.43.3-microsoft-standard-WSL2|
|CPUs|Intel(R) Core(TM) i7-8700 CPU @ 3.20GHz (12 x 3192)|
|Memory (System)|15.56GB (6.42GB free)|
|VM|0%|
</details><details>
<summary>A/B Experiments</summary>

```
vsliv695:30137379
vsins829:30139715
vsliv368cf:30146710
vsreu685:30147344
python383cf:30185419
vspor879:30202332
vspor708:30202333
vspor363:30204092
pythonvspyt639:30291487
pythontb:30258533
vspre833:30321513
vshan820:30294714
vscus158:30321503
pythonvsuse255:30319630
3f5j7703:30294757
vscod805cf:30301675
vscextlangct:30310089
vsccppwtct:30312693
bridge0708:30335490

```

</details>

### PR DESCRIPTION
**Pull request by philliphoff**
Currently, the extension arbitrarily decides that the first bound port of the `tye` process found is the dashboard API endpoint.  However, if ingress is enabled, the `tye` process may bind additional ports.  If those ports are found *before* the actual dashboard API endpoint, the extension will fail to show the application.

This change has the Tye process provider check each port associated with a `tye` process to verify one of them as the dashboard API endpoint (by calling the "get endpoints" endpoint).

Resolves #125.

### CODE REVIEW

File: src/services/tyeProcessProvider.ts [hunk 1]
```diff
@@ -60,26 +79,48 @@ export default class LocalTyeProcessProvider implements TyeProcessProvider {
...
-    private async getPortForProcess(pid: number): Promise<ProspectiveTyeProcess> {
+    private async getTyeProcessWithPorts(pid: number): Promise<TyeProcessWithPorts> {
         const ports = await this.portProvider.getPortsForProcess(pid);
 
-        return { pid, dashboardPort: ports[0] }
+        return { pid, ports };
     }
-}
 
+    private async getTyeProcessFromPorts(process: TyeProcessWithPorts): Promise<TyeProcess | undefined> {
```
Comment (karolz-ms) [inline src/services/tyeProcessProvider.ts:105]: `determineDashboardPort` ?

The `getTyeProcessWithPorts` vs `getTyeProcessFromPorts` nomenclature is a little confusing...

File: src/services/tyeProcessProvider.ts [hunk 2]
```diff
@@ -60,26 +79,48 @@ export default class LocalTyeProcessProvider implements TyeProcessProvider {
...
+        for (const port of process.ports) {
+            const dashboardUri = vscode.Uri.parse(`http://localhost:${port}`);
+            const client = this.tyeClientProvider(dashboardUri);
+
+            if (client) {
+                try {
+                    const endpoints = await client.getEndpoints();
+
+                    if (endpoints) {
```
Comment (karolz-ms) [inline src/services/tyeProcessProvider.ts:114]: Consider adding a comment that better verification of the dashboard endpoint is planned. In a microservice app there might be many things that respond to `api/v1` path ☺️

**Review by karolz-ms — APPROVED** [review]

**Review by pratiksanglikar — APPROVED** [review]

File: src/services/tyeProcessProvider.ts (hunk 2 above)
Comment (philliphoff) [reply src/services/tyeProcessProvider.ts:114]: Yes, perhaps we could add a more "uniquely Tye" endpoint to the dashboard.  (I need to dust off my mDNS PR for Tye, which would sidestep the issue in that Tye itself could advertise its dashboard port, but haven't had time.)

Comment (philliphoff) [reply src/services/tyeProcessProvider.ts:114]: Added TODO.

File: src/services/tyeProcessProvider.ts (hunk 1 above)
Comment (philliphoff) [reply src/services/tyeProcessProvider.ts:105]: Yes, the difference was a little subtle.  Went with your suggestion.

**Pull request merged**